### PR TITLE
Fastdds type support extensions

### PR DESCRIPTION
**Pull request by MiguelCompany**
This extends the initialization callback `max_serialized_size` so it returns whether the type support refers to a POD type.

This change is required in order to support zero-copy on rmw_fastrtps_cpp.

Connects to ros2/rmw_fastrtps#516
Connects to ros2/rmw_connextdds#14

### CODE REVIEW

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 1]
```diff
@@ -588,10 +594,15 @@ if isinstance(type_, AbstractNestedType):
   return current_alignment - initial_alignment;
 }
 
-static size_t _@(message.structure.namespaced_type.name)__max_serialized_size(bool & full_bounded)
+static size_t _@(message.structure.namespaced_type.name)__max_serialized_size(bool & full_bounded, bool & is_plain)
 {
+  // Start considering the type is plain.
+  // Internal methods will set values to false when necessary.
+  full_bounded = true;
+  is_plain = true;
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:602]: @MiguelCompany nit: consider moving this into `max_serialized_size_*()`. It's unusual for a function to not have a well defined output argument value.

File: rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em [hunk 2]
```diff
@@ -456,9 +462,14 @@ static uint32_t _@(message.structure.namespaced_type.name)__get_serialized_size(
...
 }
 
-static size_t _@(message.structure.namespaced_type.name)__max_serialized_size(bool & full_bounded)
+static size_t _@(message.structure.namespaced_type.name)__max_serialized_size(bool & full_bounded, bool & is_plain)
 {
-  return max_serialized_size_@(message.structure.namespaced_type.name)(full_bounded, 0);
+  // Start considering the type is plain.
+  // Internal methods will set values to false when necessary.
+  full_bounded = true;
+  is_plain = true;
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em:470]: @MiguelCompany same as above.

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 1 above)
Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:602]: I understand your point. However, as the internal methods may be called from other internal methods (in the case of composed types), a parent may have turned one of the booleans to `false`.

Consider the following messages:

```
# Inner.msg
uint32 index
uin8[10] data

# Outer.msg
string title
Inner data
```

When calling `_Outer__max_serialized_size`, the following sequence of events will occur:
* `_Outer__max_serialized_size` will set both booleans to `true` and call `max_serialized_size_Outer`
* `max_serialized_size_Outer` will set both booleans to `false` due to the string `title`
* `max_serialized_size_Outer` will call `max_serialized_size_Inner` due to the field `data`
* `max_serialized_size_Inner` will not change the booleans

When calling `_Inner__max_serialized_size`, the following sequence of events will occur:
* `_Inner__max_serialized_size` will set both booleans to `true` and call `max_serialized_size_Inner`
* `max_serialized_size_Inner` will not change the booleans

In summary:
* The public API on this package is the one defined for the callbacks.
* For that API, this PR is fixing the code in order to always have a well-defined output argument value.
* `max_serialized_size_*` should be considered internal, but needs to be exported in order for messages to be included on other messages

File: rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em (hunk 2 above)
Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em:470]: See above

File: rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h [hunk 3]
```diff
@@ -19,6 +19,9 @@
 
 #include <fastcdr/Cdr.h>
 
+/// Feature define to allow API version detection
+#define ROSIDL_TYPESUPPORT_FASTRTPS_HAS_PLAIN_TYPES
```
Comment (asorbini) [inline rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h:23]: Have you considered defining it as:

```cpp
#ifndef ROSIDL_TYPESUPPORT_FASTRTPS_HAS_PLAIN_TYPES
#define ROSIDL_TYPESUPPORT_FASTRTPS_HAS_PLAIN_TYPES    1
#endif /* ROSIDL_TYPESUPPORT_FASTRTPS_HAS_PLAIN_TYPES */
```

and then test it with `#if ROSIDL_TYPESUPPORT_FASTRTPS_HAS_PLAIN_TYPES` instead of `#ifdef`?

I think the guard with `#if` is a bit more flexible, because you can still define the flag to be 0 to disable it, and having the definition guarded by `#ifndef` will allow you to override the flag from command-line/cmake if needed (i.e. `gcc ... -D<FLAG>=0|1`).

Just a suggestion, maybe I'm overlooking something.

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 1 above)
Comment (hidmic) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:602]: Hmm, that's a problem only because `max_serialized_size_*()` functions for composed types are given the same `is_plain` to mutate. I won't block on this, _maybe_ it's best for performance, but I will point out it obscures the logic a bit. A logical AND of the `is_plain` variable for each field in a type is easier to read.

File: rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h (hunk 3 above)
Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h:23]: It is not a configuration define. I mean the purpose is not to decide on build time which API to offer, but to inform users of the API that `max_serialized_size` has the `is_plain` argument.

It is kind of an API version define.

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 1 above)
Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: Ok. Fixed in e69b64e0ead2b612271ada71ca52fa2473d1d9cd

File: rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em (hunk 2 above)
Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em]: Fixed in e69b64e0ead2b612271ada71ca52fa2473d1d9cd

**Comment by MiguelCompany** [comment]
@hidmic I addressed your concerns

@clalancette May we run a CI on this?

**Review by hidmic — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14308/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9098/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11994/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14413/)

**Comment by MiguelCompany** [comment]
@clalancette I think you forgot to include ros2/rmw_connextdds#14

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14317/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9106/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12003/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14422/)

**Comment by MiguelCompany** [comment]
@clalancette As stated in the description, this PR

> Connects to [ros2/rmw_fastrtps#516](https://github.com/ros2/rmw_fastrtps/pull/516)
> Connects to [ros2/rmw_connextdds#14](https://github.com/ros2/rmw_connextdds/pull/14)

So both should be used on the build. Sorry for the confusion.

Thank you!

**Comment by MiguelCompany** [comment]
@clalancette Friendly ping

**Comment by MiguelCompany** [comment]
@clalancette I rebased this and added commit 7e709c25444d01763231f4564c8f201098b13144, in which I make the change binary compatible.

**Comment by clalancette** [comment]
Here's a CI with what should be all of the correct pieces in place:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14421/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9194/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12095/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14521/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/14522/)

**Comment by MiguelCompany** [comment]
@clalancette I just fixed the linter errors on the generated code.

**Comment by clalancette** [comment]
Here's another try:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14432/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9204/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12104/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14532/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/14533/)

**Comment by clalancette** [comment]
CI again:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14445/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9216/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12116/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14548/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/14549/)

**Comment by MiguelCompany** [comment]
@clalancette eProsima/Fast-DDS#1931 fixes the assertions

**Comment by clalancette** [comment]
Windows Debug again: [Build Status](http://ci.ros2.org/job/ci_windows/14563/)

**Comment by MiguelCompany** [comment]
> Windows Debug again: [Build Status](http://ci.ros2.org/job/ci_windows/14563/)

@clalancette The has been stalled for three days before even starting to build.

```
15:38:00 C:\J\workspace\ci_windows>rem Kill any running docker containers, which may be leftover from aborted jobs 
15:38:00 
15:38:00 C:\J\workspace\ci_windows>powershell -Command "if ($(docker ps -q) -ne $null) { docker stop $(docker ps -q)}" 
```

**Comment by clalancette** [comment]
Let's try again:
Windows Debug again: [Build Status](http://ci.ros2.org/job/ci_windows/14580/)

File: rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h [hunk 4]
```diff
@@ -19,6 +19,13 @@
...
+
+#define ROSIDL_TYPESUPPORT_FASTRTPS_UNBOUNDED_TYPE 0x00
+#define ROSIDL_TYPESUPPORT_FASTRTPS_BOUNDED_TYPE   0x01
+#define ROSIDL_TYPESUPPORT_FASTRTPS_PLAIN_TYPE     0x03
```
Comment (hidmic) [inline rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/message_type_support.h:27]: @MiguelCompany meta: why not an `enum`? You'd get type checking.

**Comment by hidmic** [comment]
@clalancette it looks like this PR's ready to go?

**Comment by clalancette** [comment]
> @clalancette it looks like this PR's ready to go?

Yeah, I think this is ready to go in with https://github.com/ros2/rmw_fastrtps/pull/523 and https://github.com/ros2/rmw_connextdds/pull/14 , but I'd like to run one more CI on it.  But I'm going to hold off on that until we sort out https://github.com/ros2/rclcpp/pull/1657

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14574/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9330/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12237/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14700/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/14701/)

**Comment by clalancette** [comment]
The warnings are all just an artifact of this needing to be rebased.  The test failures look like known flakes.  So I'm going to go ahead and merge this along with the rest of the PRs for this.
